### PR TITLE
Add support for 'in' operator

### LINKED ISSUE #17: Add support for the in (a list) operator
Checks whether there is an element in the array equal to the specified value
``` javascript
{
    "resourceType": "Microsoft.Compute/virtualMachines",
    "path": "apiVersion",
    "in": [
         "2020-06-01",
         "2019-12-01",
         "2019-07-01",
         "2019-03-01"
    ] // Evaluates to `true` because the value of "apiVersion" in the template ("2020-06-01") is equal to one of the values in the array specified for `in`.
}
```

### PR DESCRIPTION
**Pull request by VeraBE**
Closes #17

### CODE REVIEW

File: docs/authoring-json-rules.md [hunk 1]
```diff
@@ -222,10 +222,10 @@ Example:
 }
 ```
 
-#### **In [[*]](#note)**
+#### **In**
 *Type: array of basic JSON values (integer, float, string, bool, null)*
 
-Evaluates the value of the `path` in the template using the `equals` operator for each value specified in the array.  If any results in `true`, `in` will evaluate to true; `false` otherwise.  All values in the array must be of the same type.
+Evaluates the value of the `path` in the template using the `equals` operator for each value specified in the array.  If any results in `true`, `in` will evaluate to true; `false` otherwise.
```
Comment (JohnathonMohr) [inline docs/authoring-json-rules.md:228]: It would be helpful to add a small statement that different types within the array *are* allowed.

File: src/Analyzer.JsonRuleEngine.UnitTests/JsonRuleEngineTestsUtilities.cs [hunk 2]
```diff
@@ -0,0 +1,18 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Azure.Templates.Analyzer.RuleEngines.JsonEngine.UnitTests
+{
+    public class JsonRuleEngineTestsUtilities
```
Comment (JohnathonMohr) [inline src/Analyzer.JsonRuleEngine.UnitTests/JsonRuleEngineTestsUtilities.cs:9]: Rename class and file to TestUtilities.

File: src/Analyzer.JsonRuleEngine/Operators/InOperator.cs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        /// <summary>
+        /// Gets the name of this operator.
+        /// </summary>
+        public override string Name => "In";
+
+        /// <summary>
+        /// Creates an InOperator.
+        /// </summary>
+        /// <param name="specifiedValue">The value specified in the JSON rule, it has to be an array of basic JSON values.</param>
+        /// <param name="isNegative">Whether the result of EvaluateExpression() should be negated or not.</param>
```
Comment (JohnathonMohr) [inline src/Analyzer.JsonRuleEngine/Operators/InOperator.cs:22]: For references to code (like EvaluateExpression() here), put it in a `<see cref="..."/>` tag.

File: src/Analyzer.JsonRuleEngine/Operators/InOperator.cs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        public InOperator(JToken specifiedValue, bool isNegative)
+        {
+            this.SpecifiedValue = specifiedValue;
```
Comment (JohnathonMohr) [inline src/Analyzer.JsonRuleEngine/Operators/InOperator.cs:25]: There should be validation somewhere that what's passed in is actually a JArray.  The earlier this validation can happen in the runtime, the better.  If it can be enforced as part of the schema, that would be best.

File: src/Analyzer.JsonRuleEngine/Operators/InOperator.cs [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+            foreach (var arrayElement in this.SpecifiedValue)
+            {
+                if (equalsOperator.EvaluateExpression(arrayElement))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+
```
Comment (JohnathonMohr) [inline src/Analyzer.JsonRuleEngine/Operators/InOperator.cs:47]: Remove extra line

File: src/Analyzer.JsonRuleEngine.UnitTests/InOperatorTests.cs [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    {
+        [DataTestMethod]
+        [DataRow(true, new object[] { "anotherValue", "aValue" }, DisplayName = "String is in string array")]
+        [DataRow(false, new object[] { "anotherValue", "otherValue" }, DisplayName = "String is not in string array")]
+        [DataRow(false, new object[] { }, DisplayName = "String is not in empty array")]
+        [DataRow(true, new object[] { "anotherValue", 4, false, 3.5, null, "aValue" }, DisplayName = "String is in mixed array")]
+        [DataRow(false, new object[] { "anotherValue", 4, false, 3.5, null, "otherValue" }, DisplayName = "String is not in mixed array")]
+        public void EvaluateExpression(bool evaluationResult, params object[] arrayOfValues)
+        {
+            var aValue = "aValue";
```
Comment (JohnathonMohr) [inline src/Analyzer.JsonRuleEngine.UnitTests/InOperatorTests.cs:21]: Add tests to find other data types inside the array as well (besides a string).

File: src/Analyzer.JsonRuleEngine.UnitTests/InOperatorTests.cs [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+            var inOperator = new InOperator(arrayOfValuesJToken, isNegative: false);
+            Assert.AreEqual(evaluationResult, inOperator.EvaluateExpression(valueJToken));
+        }
+
+        [TestMethod]
+        public void GetName_ReturnsCorrectName()
+        {
+            Assert.AreEqual("In", new InOperator(new JObject(), false).Name);
+        }
+    }
```
Comment (JohnathonMohr) [inline src/Analyzer.JsonRuleEngine.UnitTests/InOperatorTests.cs:35]: Add tests for the constructor as well.

File: src/Analyzer.JsonRuleEngine.UnitTests/ExpressionConverterTests.cs [hunk 8]
```diff
@@ -25,6 +25,13 @@ public class ExpressionConverterTests
             .Where(property => property.Attribute != null)
             .ToDictionary(property => property.Attribute.PropertyName ?? property.Property.Name, property => property.Property, StringComparer.OrdinalIgnoreCase);
 
+        public static IReadOnlyList<object[]> InOperatorTestScenarios { get; } = new List<object[]>
```
Comment (JohnathonMohr) [inline src/Analyzer.JsonRuleEngine.UnitTests/ExpressionConverterTests.cs:28]: Add a comment above these to explain they are for the ReadJson_LeafWithValidOperator_ReturnsCorrectTypeAndValues test.

**Review by JohnathonMohr — CHANGES_REQUESTED** [review]
Looking pretty good overall.  Just a few things to add.

File: src/Analyzer.JsonRuleEngine/Operators/InOperator.cs (hunk 3 above)
Comment (VeraBE) [reply src/Analyzer.JsonRuleEngine/Operators/InOperator.cs:22]: Adding it everywhere then :)

File: src/Analyzer.JsonRuleEngine.UnitTests/InOperatorTests.cs (hunk 6 above)
Comment (VeraBE) [reply src/Analyzer.JsonRuleEngine.UnitTests/InOperatorTests.cs:21]: I thought it wasn't necessary because for the comparison I'm using the 'equals' operator that covers those cases in its own tests, do you think I should add those tests anyways?

Comment (JohnathonMohr) [reply src/Analyzer.JsonRuleEngine.UnitTests/InOperatorTests.cs:21]: Using the 'equals' operator is a good way to implement it.  Having tests ensures it's working as intended, and also protects against future changes that may inadvertently break it.

File: src/Analyzer.JsonRuleEngine/Operators/InOperator.cs (hunk 4 above)
Comment (VeraBE) [reply src/Analyzer.JsonRuleEngine/Operators/InOperator.cs:25]: I'm thinking on adding it in LeafExpressionDefinition, I couldn't find an earlier stage to make the validation, what do you think?

File: src/Analyzer.JsonRuleEngine.UnitTests/InOperatorTests.cs (hunk 7 above)
Comment (VeraBE) [reply src/Analyzer.JsonRuleEngine.UnitTests/InOperatorTests.cs:43]: I'm assuming you wrote this because of the array validation issue. I added a test for that in ExpressionConverterTests. It doesn't even compile if I try to use a non JArray as parameter in the constructor

File: src/Analyzer.JsonRuleEngine.UnitTests/ExpressionConverterTests.cs [hunk 9]
```diff
@@ -210,12 +210,15 @@ void ValidateExpression(ExpressionDefinition expression, Type expectedSpecificTy
...
-                {{
-                    ""resourceType"": ""resourceType"",
-                    ""path"": ""path"",
-                    ""{operatorProperty}"": {JsonConvert.SerializeObject(operatorValue)}
-                }}");
+            ReadSimpleOperatorJson(operatorProperty, operatorValue);
+        }
+
+        [DataTestMethod]
+        [DataRow("in", "aString", DisplayName = "\"In\": \"aString\"")]
```
Comment (JohnathonMohr) [inline src/Analyzer.JsonRuleEngine.UnitTests/ExpressionConverterTests.cs:217]: Add tests for other types as well.

File: src/Analyzer.JsonRuleEngine.UnitTests/InOperatorTests.cs [hunk 10]
```diff
@@ -39,7 +38,7 @@ public void EvaluateExpression(bool evaluationResult, object desiredValue, param
         [TestMethod]
         public void GetName_ReturnsCorrectName()
         {
-            Assert.AreEqual("In", new InOperator(new JObject(), false).Name);
+            Assert.AreEqual("In", new InOperator(new JArray(), false).Name);
         }
     }
```
Comment (JohnathonMohr) [inline src/Analyzer.JsonRuleEngine.UnitTests/InOperatorTests.cs:43]: Also add a test for evaluating a null JToken.  This is different than evaluating a JToken for which the value in it is null.
For example, something like this:
``` diff
- var desiredValueJToken = TestUtilities.ToJToken(null);
+ var desiredValueJToken = null;
inOperator.EvaluateExpression(desiredValueJToken)
```

This will probably need to be a test on its own, since it doesn't fit what other tests are doing.

**Review by JohnathonMohr — CHANGES_REQUESTED** [review]

File: src/Analyzer.JsonRuleEngine.UnitTests/InOperatorTests.cs (hunk 10 above)
Comment (VeraBE) [reply src/Analyzer.JsonRuleEngine.UnitTests/InOperatorTests.cs:43]: I can't compile `inOperator.EvaluateExpresion(null)`: _Cannot assign <null> to an implicitly typed variable_

Comment (JohnathonMohr) [reply src/Analyzer.JsonRuleEngine.UnitTests/InOperatorTests.cs:43]: Whoops - `JToken desiredValueJToken = null;` 🙂

File: src/Analyzer.JsonRuleEngine/Operators/InOperator.cs [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+        /// <param name="tokenToEvaluate">The JToken to evaluate.</param>
+        /// <returns>A value indicating whether or not the evaluation passed.</returns>
+        public override bool EvaluateExpression(JToken tokenToEvaluate)
+        {
+            if (tokenToEvaluate == null)
+            {
+                return false;
+            }
+
+            var equalsOperator = new EqualsOperator(tokenToEvaluate, isNegative: false);
```
Comment (yane3628) [inline src/Analyzer.JsonRuleEngine/Operators/InOperator.cs:41]: I believe this should pass the value `this.IsNegative` instead of `false`.

**Review by yane3628 — CHANGES_REQUESTED** [review]

File: src/Analyzer.JsonRuleEngine/Operators/InOperator.cs (hunk 11 above)
Comment (VeraBE) [reply src/Analyzer.JsonRuleEngine/Operators/InOperator.cs:41]: This operator doesn't use the isNegative attribute, it's always false, do you think I should change it anyways?

File: src/Analyzer.JsonRuleEngine.UnitTests/InOperatorTests.cs [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+        [DataRow(true, false, new object[] { 4, 3.5, "aValue", false, null }, DisplayName = "Boolean is in")]
+        [DataRow(false, true, new object[] { 4, 3.5, "aValue", false, null }, DisplayName = "Boolean is not in")]
+        [DataRow(true, null, new object[] { 4, 3.5, "aValue", false, null }, DisplayName = "Null is in")]
+        [DataRow(false, null, new object[] { 4, 3.5, "aValue", false }, DisplayName = "Null is not in")]
+        [DataRow(false, 7, new object[] { }, DisplayName = "Integer is not in empty array")]
+        [DataRow(false, 7.7, new object[] { }, DisplayName = "Float is not in empty array")]
+        [DataRow(false, "aValue", new object[] { }, DisplayName = "String is not in empty array")]
+        [DataRow(false, true, new object[] { }, DisplayName = "Boolean is not in empty array")]
+        [DataRow(false, null, new object[] { }, DisplayName = "Null is not in empty array")]
+        public void EvaluateExpression(bool evaluationResult, object desiredValue, object arrayOfValues)
```
[the remark this answers is not in the document]
Comment (VeraBE) [reply src/Analyzer.JsonRuleEngine.UnitTests/InOperatorTests.cs:29]: I see how I could do that for the null test, something like: EvaluateExpression_NullAsTokenToEvaluate_ReturnsFalse, right? But for the other test the outcome and the case being tested changes in each DataRow. It's okay if I leave that one the way it is?

File: src/Analyzer.JsonRuleEngine/Operators/InOperator.cs (hunk 11 above)
Comment (JohnathonMohr) [reply src/Analyzer.JsonRuleEngine/Operators/InOperator.cs:41]: This is what I mentioned offline.  You are setting `this.IsNegative` in the constructor, but aren't using it here.  Either don't expose it as a parameter in the constructor, or honor whatever it's set to and add tests for it.

File: src/Analyzer.JsonRuleEngine.UnitTests/InOperatorTests.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (yane3628) [reply src/Analyzer.JsonRuleEngine.UnitTests/InOperatorTests.cs:29]: You can do it in a generic way. Something like EvaluateExpression_ValidDataType_ReturnsExpectedEvaluationResult

[the remark this answers is not in the document]
Comment (JohnathonMohr) [reply src/Analyzer.JsonRuleEngine.UnitTests/InOperatorTests.cs:29]: The Equals operator tests have good examples, like `EvaluateExpression_PropertyIsEqual_EqualsExpressionIsTrue_NotEqualsExpressionIsFalse`.  It might be good to separate this test into two - one where the value is in the array, one where it isn't.

**Review by JohnathonMohr — APPROVED** [review]

**Review by yane3628 — APPROVED** [review]
